Declarative diffusion config (#1319)
C2SM/icon4py: docs/development/ADRs/0002_configuration_classes.md

---
tags: [config, fortran, icon]
---

# [Configuration Classes]

- **Status**: valid
- **Authors**: Rico Häuselmann (@DropD)
- **Created**: 2026-06-17
- **Updated**: 2026-06-17

> (Wh)Y-statement: In the context of [use case/user story u], facing [concern c] we decided for [option o] and neglected [other options], to achieve [system qualities/desired consequences], accepting [downside/undesired consequences].

Reviewer: should this line be deleteed?

Author: Yes, thanks!

While making Icon4Py user-configurable, facing code duplication between modules, we decided to provide a standard way to declaratively encode common attributes of configuration options alongside utilities that can then be reused across modules.

## Context

> We notice that currently, each module has a class, which contains it's configuration. Each of these classes knows how to construct itself from a dictionary that represents an ICON namelist. The mapping is contained implicitly in a classmethod. Usually the class attributes that have an equivalent in ICON are annotated with documentation comments (where to find in ICON namelists and what it does).

Reviewer: ```suggestion
We notice that, currently, each module has a class which contains its configuration. Each of these classes knows how to construct itself from a dictionary that represents an ICON namelist. The mapping is contained implicitly in a classmethod. Usually the class attributes that have an equivalent in ICON are annotated with documentation comments (where to find in ICON namelists and what it does).
```

Author: Thanks, I will correct the `it's` typo (I checked and none of the commas in either of our versions is actually appropriate here).

> Really, given the mapping for each configuration option in each of those classes, the logic is the same for all of them.

Reviewer: ```suggestion
Given the mapping for each configuration option in each of those classes, the logic is the same for all of them.
```

Extrapolating from this case of duplication we find the same for:

- describing configuration options to users (currently only documentation comments exist for this)
- validation (currently partially done in monolithic, imperative validation methods)

## Decision

> The chosen pattern is for configuration classes to look as Follows

Reviewer: ```suggestion
The chosen pattern is for configuration classes to look as follows
```

```python
import dataclasses
import typing

from icon4py.model.common.config import options as common_conf_opt

@dataclasses.dataclass
class MyModuleConfig:
    some_configuration_option: typing.Annotated[
        int,
        common_conf_opt.ConfigOption(

>             description="This is help telling the user what `some_configuration_option` does.",

Reviewer: ```suggestion
            description="Describe what `some_configuration_option` does.",
```

            icon_equivalent=common_conf_opt.IconOption(
                name="isomcfgop",
                path=("diffusion_nml",),  # for nested sections: ("toplevel_nml", "sublevel_nml")
                read_from_icon=True,  # optional (default: True), whether to actually read this from ICON namelists
                list_to_value=False,  # optional (default: False), whether to map a list of values in ICON namelists to a single value

>             ),

Reviewer: I love this, especially the **very**, **very** realistic icon variable name 😍

        ),
    ] = 4  # default value
```

The pattern is meant to be extended in the future as the need arises. One example would be validation:

```python
def validate_some_configuration_option_with_some_flag(
    some_configuration_option: int, other_option: bool
) -> None:
    if not other_option:
        raise ConfigValidationError(
            f"'some_configuration_option' should only be set if 'other_option' is True."
        )

...

@dataclasses.dataclass
class MyModuleConfig:
    some_configuration_option: typing.Annotated[
        int,
        ConfigOption(
            ...,
            validators=(
                DependsOn("some_flag", condition=validate_some_configuration_option_with_some_flag),
                Min(0),
                Max(59000),
                ...,
            ),
        ),
    ] = 4

> ```

Reviewer: This is also very nice

Author: This is basically copied from Pydantic / Django / SqlAlchemy etc, no need to reinvent the wheel.

## Consequences

Code for mapping from ICON inputs to Icon4Py configurations can be shared between modules.

> It becomes possible to automatically provide helpful (for users) information for all available configuration options (for a given set of modules).

Reviewer: ```suggestion
It becomes possible to automatically provide helpful information for all available configuration options.
```

> Different user facing utilities for configuring / inspecting Icon4Py runs will not have to duplicate logic for

Reviewer: ```suggestion
Different user facing utilities for configuring / inspecting ICON4Py runs will not have to duplicate logic for
```

Author: Replaced throughout

- mapping to / from ICON (fortran)
- validation
- documentation
- etc

> The current solution will not achieve maximum separation isolation between Icon4Py internals and their mapping to and from ICON concepts. However it provides a structured path towards it.

Reviewer: ```suggestion
The current solution will not achieve maximum separation between ICON4Py internals and their mapping to and from ICON concepts. However, it provides a structured path towards it.
```

> Another downside is a certain loss of type checking when using the provided utilities. This can be made up for in tests.

Reviewer: can you specify what you mean with "a certain loss of type checking" and / or why?

Author: Added an example code snippet

## Alternatives Considered

### Separating (fortran-) ICON mapping information from the rest

> - Pro: allow free evolution of Icon4Py internals, decoupled from ICON

Reviewer: ```suggestion
- Pro: allows for free evolution of ICON4Py internals, decoupled from ICON
```

- Con: more complicated, less (for now) helpful information immediately available

> This could be achieved, however, by simply moving all the instances of `IconOption` out of `ConfigOption` annotations and into a separate (per-module) dictionary, keyed by the name (or class AND name) of the option it belongs to.

Reviewer: ```suggestion
This could be achieved, however, by moving all the instances of `IconOption` out of `ConfigOption` annotations and into a separate (per-module) dictionary, keyed by the name (or class AND name) of the option it belongs to.
```

    DiffusionConfig, {"diffusion_nml": {"itype_vn_diffu": 1}}
)  # no static type checking can be performed here, because the mapping logic is run-time only
```

As opposed to hand-mapping, where type checkers can infer the required types at least to some degree.

```python
# More likely usage, where static type checking is not a concern
icon_config: dict[str, Any] = read_namelists()
diffusion_config = construct_config_from_icon(icon_config)
```

## Alternatives Considered

### Separating (fortran-) ICON mapping information from the rest

- Pro: allow for free evolution of ICON4Py internals, decoupled from ICON
- Con: more complicated, less (for now) helpful information immediately available

This could be achieved, however, by moving all the instances of `IconOption` out of `ConfigOption` annotations and into a separate (per-module) dictionary, keyed by the name (or class AND name) of the option it belongs to.

Example:

```python
class DiffusionConfig:
    diffusion_type: Annotated[DiffusionType, ConfigOption(description="...", validation=...)]
    ...
```

And in `icon4py.model.atmosphere.diffusion.fortran_icon_mappings`:

```python

> CONFIC_OPTIONS: {DiffusionConfig: {"diffusion_type": ...}}

Reviewer 2: ```suggestion
CONFIG_OPTIONS: {DiffusionConfig: {"diffusion_type": ...}}
```

```

To arrive at the same usability, `icon4py.common.fortan_icon_mappings` could contain some infrastructure to find the relevant mapping for each configuration class (via module path heuristics or entry points or similar), so that a different version of `icon4py.common.config.options.construct_config_from_icon` could be provided.

> 

Reviewer 2: I like this section. I think the long term goal should still be to fully separate the python <-> fortran mapping and you might want to explicitly mention that here or together with

> The current solution will not achieve maximum separation between internals and their mapping to and from ICON concepts. However it provides a structured path towards it.

to make it clear to someone reading this that this is not the final shape of the python <-> fortran mapping.

Author: I was going to put this into a separate ADR, so it can be more broadly reviewed without blocking any code changes.

Author: I rephrased the section to make it apply more specifically to configuration classes and to state that the decision was only on whether to separate more now or later.
